fix: cp-13.10.0 automatically restart the extension after MM is updated in order to work around the chromium bug https://issues.chromium.org/issues/40805401

davidmurdoch opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

Automatically restart the extension after MM is updated in order to work around this chromium bug: https://issues.chromium.org/issues/40805401

It works by listening for the `onInstalled` event, and if it the `event.reason` is `"update"`  and the `event.previousVersion` is NOT the current version, and the `event.previousVersion` has _not_ already been recorded in the `AppStateController`, we will reload the extension.

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/37552?quickstart=1)

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and accurately)
-->

CHANGELOG entry: Automatically restart the extension after MM is updated in order to work around this chromium bug: https://issues.chromium.org/issues/40805401

## **Related issues**

Fixes: #37503 (by hopefully making it unnecessary!)

## **Manual testing steps**

### Scenarios:

#### Doesn't fire on initial installation

1. Install a version of the extension that does NOT have lavamoat (we need to use `globalThis.chrome`) and onboard
2. open the service worker/background console
3. logs starting with "[onUpdate]:" should not appear

#### Doesn't reload before onboarding

1. close your browser completely (any and all Devtools panels too!)
2. re-open your browser
3. open the service_worker/background console
4. simulate an update event by running this code in the console: `chrome.runtime.onInstalled.dispatch({reason:"update", "previousVersion": "13.10.9"});`
5. should should see: `[onUpdate]: Should reload: undefined` (`undefined` indicates we didn't fetch our remote feature flags) in the logs

#### Doesn't fetch feature flags until after onboarding (`Basic functionality` is left enabled during onboarding)

1. Monitor the Network tab for `https://client-config.api.cx.metamask.io/v1/flags`
2. Onboard
3. you shouldn't see https://client-config.api.cx.metamask.io/v1/flags until AFTER onboarding completes.

_note: you can ensure we _don't_ fetch feature flags after onboarding by toggling `Basic functionality` to off as well._

#### Reloads after onboarding (`Basic functionality` is on)

1. close your browser completely (any and all Devtools panels too!)
2. re-open your browser
3. open the service_worker/background console
4. simulate an update event by running this code in the console: `chrome.runtime.onInstalled.dispatch({reason:"update", "previousVersion": "13.10.10"});`
5. it will reload only _once_.

#### RPC Fallback still works

1. download a build from this PR (the build needs Quicknode ENV vars, which you probably don't have)
2. install and onboard
3. prevent connections to `mainnet.infura.io` however you'd like (probably via `/etc/hosts`)
4. Check the `service_worker` console and ensure quicknode URLs show up when you open the UI.

<!-- 
## **Screenshots/Recordings**


### **Before**


### **After**

 -->
## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.



<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds a safe-reload flow on extension update (Chromium, gated by a remote flag), defers/upgrades remote feature flag handling, and wires flags into NetworkController to enable/disable RPC failover; updates messengers, tests, and fixtures.
> 
> - **Background**:
>   - Add `onUpdate` handler to record update metadata and trigger safe reload on Chromium when `extensionPlatformAutoReloadAfterUpdate` flag is true.
>   - Handle `runtime.onInstalled` update path asynchronously and invoke `onUpdate(controller, platform, previousVersion, requestSafeReload)`.
>   - Lazily fetch remote feature flags on UI open via `updateRemoteFeatureFlags` util.
>   - Simplify `setupController` signature (stop passing `requestSafeReload`).
> - **Remote Feature Flags**:
>   - Initialize `RemoteFeatureFlagController` with persisted state; gate enablement by onboarding completion and `useExternalServices`.
>   - Subscribe to `PreferencesController` and `OnboardingController` state changes to enable/disable and refresh flags.
>   - Extend messengers to allow `RemoteFeatureFlagController:getState` and related `stateChange` events.
> - **Network**:
>   - Use remote flag `walletFrameworkRpcFailoverEnabled` to set `NetworkController` `isRpcFailoverEnabled` at init and toggle on flag changes.
> - **Tests/Fixtures**:
>   - Update unit tests to cover new messenger actions and RPC failover gating.
>   - Add `FixtureBuilder.withRemoteFeatureFlags` and set flags/manifest in e2e tests.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 802612e512cbc6a8b73b7263a22cd5643feda065. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

davidmurdoch committed:
fix: automatically restart the extension after MM is updated in order to work around the chromium bug https://issues.chromium.org/issues/40805401

davidmurdoch committed:
lint

davidmurdoch committed:
.

davidmurdoch committed:
add safe guard and comments

davidmurdoch commented on app/scripts/background.js:
```diff
@@ -1739,6 +1737,15 @@ async function onUpdate(previousVersion) {
   controller.appStateController.setLastUpdatedAt(Date.now());
   if (previousVersion) {
     controller.appStateController.setLastUpdatedFromVersion(previousVersion);
```
I've made a changes in order to delay the `requestSafeReload` call (delayed twice, once by store.update, and once by setImmediate)

I also guard against a reload loop by checking the previous value of `lastUpdatedFromVersion`, ensuring it has changed.

Thoughts?

You can test it by running

`chrome.runtime.onInstalled.dispatch({reason:"update", "previousVersion": "1.2.3.4"});`

in the background (with lavamoat disabled).

davidmurdoch committed:
lint

davidmurdoch committed:
paranoia

davidmurdoch committed:
persist remote feature flags

davidmurdoch committed:
persist remote feature flags

davidmurdoch committed:
persist remote feature flags

davidmurdoch committed:
persist remote feature flags

davidmurdoch committed:
persist remote feature flags

davidmurdoch committed:
persist remote feature flags

davidmurdoch committed:
persist remote feature flags

davidmurdoch committed:
add new tests and quiknode.pro

davidmurdoch committed:
remove problematic test

davidmurdoch committed:
remove other test

davidmurdoch committed:
_

davidmurdoch committed:
Merge branch 'main' into restart

davidmurdoch committed:
lint

davidmurdoch committed:
Update privacy-snapshot.json

davidmurdoch committed:
clean up and refactor

davidmurdoch committed:
Merge branch 'restart' of https://github.com/MetaMask/metamask-extension into restart

davidmurdoch committed:
Merge branch 'main' into restart

davidmurdoch committed:
remove comment

davidmurdoch committed:
Merge branch 'main' into restart

davidmurdoch committed:
try this

davidmurdoch committed:
lint

davidmurdoch committed:
remove a stray copy paste

davidmurdoch committed:
move to ts

davidmurdoch commented on app/scripts/background.js:
```diff
@@ -1704,35 +1695,6 @@ if (
- *
- * @param previousVersion
- */
-async function onUpdate(previousVersion) {
```
I moved this to a typescript file

davidmurdoch commented on app/scripts/controller-init/remote-feature-flag-controller-init.ts:
```diff
@@ -58,20 +58,35 @@ export function getConfigForRemoteFeatureFlagRequest() {
     messenger: controllerMessenger,
     fetchInterval: 15 * 60 * 1000, // 15 minutes in milliseconds
-    disabled: !preferencesState.useExternalServices,
+    disabled: getIsDisabled(),
```
this is now the result of the combination of two controller states

davidmurdoch commented on app/scripts/onUpdate.ts:
```diff
@@ -0,0 +1,119 @@
+    // safe than better safe than... rebooting forever. :-)
+    const { remoteFeatureFlagController } = controller;
+    const shouldReload = remoteFeatureFlagController.state.remoteFeatureFlags
+      .extensionPlatformAutoReloadAfterUpdate as boolean | undefined;
```
`RemoteFeatureFlagControllerState` is:
```
type RemoteFeatureFlagControllerState = {
    remoteFeatureFlags: FeatureFlags;
    cacheTimestamp: number;
}
```

`remoteFeatureFlags` isn't optional, so it should always be there.

davidmurdoch committed:
lint

davidmurdoch committed:
woops!

Gudahtt commented on app/scripts/background.js:
```diff
@@ -1591,6 +1576,11 @@ export function setupController(
     controller.rejectAllPendingApprovals();
   }
+
+  updateNetworkControllerRpcFailoverFromRemoteFeatureFlagController(
```
Hmm. This seems a bit out-of-place here. Really we could run this check for whether RPC Failover is enabled at any point after the controller is setup I think? Seems like a better fit for the controller init module? That would avoid this pattern of passing the controller as a parameter as well, which we've been trying to avoid.

davidmurdoch replied:
Yeah, this was before I asked you about the anti-anti-pattern. I'll change it to that.

Gudahtt commented on app/scripts/background.js:
```diff
@@ -1591,6 +1576,11 @@ export function setupController(
+  updateNetworkControllerRpcFailoverFromRemoteFeatureFlagController(
+    controller.networkController,
+    controller.remoteFeatureFlagController,
+  );
```
>there's no mechanism to re-evaluate the RPC failover setting after flags are loaded

Yes there is. It's in the controller init module for the network controller.

Gudahtt commented on app/scripts/onUpdate.ts:
```diff
@@ -0,0 +1,120 @@
+    }
+  }
+  appStateController.setLastUpdatedAt(lastUpdatedAt);
+  appStateController.setLastUpdatedFromVersion(previousVersion);
```
I couldn't make sense of this comment.

davidmurdoch replied:
Yeah, its just saying if reload is called, but feature flags aren't ready yet, we'll won't get to try to `reload` for this specific version ever again.

Which is fine. Technically we can receive the event twice: when we reload the extension by calling `chrome.runtime.reload()` the `chrome.runtime.onInstalled` even is fired with `reason: "update"` _again_! But since the installed version hasn't changed (we check `previousVersion === platform.getVersion()` in `background.js` before calling this `onUpdate` function) this code doesn't end up running anyway.

davidmurdoch committed:
Merge branch 'main' into restart

davidmurdoch committed:
simplify updateRemoteFeatureFlags

Gudahtt commented on app/scripts/onUpdate.ts:
```diff
@@ -0,0 +1,120 @@
+    const appStateControllerState =
+      newState.AppStateController as AppStateControllerState;
+    if (
+      appStateControllerState.lastUpdatedFromVersion === previousVersion &&
```
... I just realized what state we're using. I thought we were going to use `appMetadataController.previousAppVersion`, I didn't realize the same state was also tracked in `appStateController` (why is it duplicated 🤔 )

The `appMetadataController` gets updated automatically upon construction, so that would save us having to wait for the next state update event. We could check it immediately here.

davidmurdoch replied:
Yeah, I was just using what was already there (as you pointed out in the slack thread). Not sure what do to going forward.

davidmurdoch committed:
use initMessenger

davidmurdoch committed:
add RemoteFeatureFlagController:getState

davidmurdoch commented on app/scripts/metamask-controller.js:
```diff
@@ -602,6 +601,7 @@ export default class MetamaskController extends EventEmitter {
       PPOMController: PPOMControllerInit,
       PhishingController: PhishingControllerInit,
       OnboardingController: OnboardingControllerInit,
+      RemoteFeatureFlagController: RemoteFeatureFlagControllerInit,
```
oh, well this is fun

davidmurdoch replied:
got it!

davidmurdoch committed:
maybe this is better?

davidmurdoch committed:
lint

davidmurdoch committed:
Merge branch 'main' into restart

davidmurdoch committed:
unit test fix

davidmurdoch committed:
i don't know

davidmurdoch committed:
sdfsdf

davidmurdoch committed:
update privacy snapshot

davidmurdoch committed:
Merge branch 'main' into restart

davidmurdoch committed:
dur

smgv committed:
fix: add missing metrics to import SRP flow and grouped Import SRP events (#37567)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

In this PR, added missing track event for SRP import: click event in Add
Wallet modal and completion event with hd_entropy_index on import.

- Metrics

  - Import Secret Recovery Phrase Clicked:  
        This event needs to be added
It should fire when the user clicks the import wallet button on the
account list.

  - Import Secret Recovery Phrase Continue Button Clicked:  
It should fire when the user finishes importing an SRP (they click
continue)
        
Also we've enhanced our tracking by grouping the Import Secret Recovery
Phrase event with specific status properties: 'started',
'continue_button_clicked', and 'completed', along with location data.
This provides visibility into the full workflow timeline from initiation
to completion.

Jira Link: https://consensyssoftware.atlassian.net/browse/SL-277

Segment Schema PR:
https://github.com/Consensys/segment-schema/pull/367/files

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/37567?quickstart=1)

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the
CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing
description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and
accurately)
-->

CHANGELOG entry: added missing metrics to import SRP flow

## **Related issues**

Fixes:

## **Manual testing steps**

1. Open extesnion.
2. Create Wallet.
3. Add wallet and track event when user clicks on 'Import a wallet' and
when user successfully import the SRP.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

<img width="1728" height="858" alt="Screenshot 2025-11-06 at 3 20 25 PM"
src="https://github.com/user-attachments/assets/ce66954e-ebf5-40e8-903b-2b219ec5d968"
/>
<img width="1727" height="1070" alt="Screenshot 2025-11-06 at 3 21
04 PM"
src="https://github.com/user-attachments/assets/aed9cbf7-3c88-4033-a4b1-31b539ceb796"
/>


## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
Extension Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.




<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Unifies SRP import tracking into a single event with
start/continue/completed statuses and adds hd_entropy_index and
discovered account counts on completion.
> 
> - **Metrics**:
> - Consolidate `ImportSecretRecoveryPhraseClicked`/`Completed` into
`ImportSecretRecoveryPhrase` with status-based phases (`started`,
`continue_button_clicked`, `completed`).
> - **Background** (`app/scripts/metamask-controller.js`):
> - On SRP import completion, track `ImportSecretRecoveryPhrase` with
`hd_entropy_index`, `number_of_solana_accounts_discovered`, and
`number_of_bitcoin_accounts_discovered`.
> - **UI**:
> -
`ui/components/multichain-accounts/add-wallet-modal/add-wallet-modal.tsx`:
Track `ImportSecretRecoveryPhrase` with `status: 'started'` when
selecting "Import a wallet" (location: Add Wallet Modal).
> - `ui/components/multichain/account-menu/account-menu.tsx`: Track
`ImportSecretRecoveryPhrase` with `status: 'started'` when clicking SRP
import (location: Account Menu).
> - `ui/pages/multi-srp/import-srp/import-srp.tsx`: Track
`ImportSecretRecoveryPhrase` with `status: 'continue_button_clicked'` on
continue (location: Multi SRP Import).
> 
> <sup>Written by [Cursor
Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit
b1ca69dc9e726d0dd3818a04cb3a0aedc3d3c17b. This will update automatically
on new commits. Configure
[here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

davidmurdoch committed:
fix test and rename a file

davidmurdoch committed:
dsaflikjhsdaflkjhasfasdf

davidmurdoch committed:
Merge branch 'main' into restart

davidmurdoch committed:
fix types

davidmurdoch committed:
Merge branch 'restart' of https://github.com/MetaMask/metamask-extension into restart

davidmurdoch commented on app/scripts/on-update.ts:
```diff
@@ -0,0 +1,87 @@
+    const { remoteFeatureFlagController } = controller;
+    const shouldReload = remoteFeatureFlagController.state.remoteFeatureFlags
+      .extensionPlatformAutoReloadAfterUpdate as boolean | undefined;
+    log.info(`[onUpdate]: Should reload: ${shouldReload}`);
```
yeah, that's fine.

davidmurdoch committed:
Merge branch 'main' into restart

davidmurdoch committed:
interesting

davidmurdoch committed:
Merge branch 'main' into restart

davidmurdoch committed:
Update privacy-snapshot.json

davidmurdoch commented on privacy-snapshot.json:
```diff
@@ -119,5 +119,6 @@
   "verify.walletconnect.com",
-  "www.4byte.directory"
+  "www.4byte.directory",
+  "rewards.uat-api.cx.metamask.io"
```
mark figured it out and fixed it in another pr

davidmurdoch committed:
Merge branch 'main' into restart

davidmurdoch committed:
well this is not good

Gudahtt committed:
Add fixture builder method to modify feature flags, and fix account syncing balance update test

davidmurdoch commented on test/e2e/tests/privacy/basic-functionality.spec.ts:
```diff
@@ -106,6 +106,13 @@ describe('MetaMask onboarding', function () {
+            sendRedesign: {
+              enabled: false,
+            },
+          }
```
```suggestion
          },
```

Gudahtt committed:
Fix lint error

Co-authored-by: David Murdoch <187813+davidmurdoch@users.noreply.github.com>

Gudahtt reviewed (approved):
LGTM!

MajorLift reviewed (approved):
LGTM
- Successfully tested on local.
- Refactors and additions all make sense.
- Increased TypeScript coverage is a great bonus.

gantunesr reviewed (approved):
Approved for Accounts CO,

- `test/e2e/tests/identity/account-syncing/balances.spec.ts`

seaona reviewed (approved):
LGTM
